client/alloc_runner_health_watcher.go

```diff
@@ -31,31 +31,29 @@ func (r *AllocRunner) watchHealth(ctx context.Context) {
 
 	// See if we should watch the allocs health
 	alloc := r.Alloc()
-	if alloc.DeploymentStatus.IsHealthy() || alloc.DeploymentStatus.IsUnhealthy() {
-		// No need to watch health as it's already set
-		return
-	}
 
-	// Neither deployments nor migrations care about system jobs so never
-	// watch their health
-	if alloc.Job.Type == structs.JobTypeSystem {
+	// Neither deployments nor migrations care about the health of
+	// non-service jobs so never watch their health
+	if alloc.Job.Type != structs.JobTypeService {
 		return
 	}
 
-	isDeploy := alloc.DeploymentID != ""
-
-	// Migrations don't consider the health of batch jobs so only watch
-	// batch health during deployments
-	if !isDeploy && alloc.Job.Type == structs.JobTypeBatch {
+	// No need to watch health as it's already set
+	if alloc.DeploymentStatus.IsHealthy() || alloc.DeploymentStatus.IsUnhealthy() {
```

Comment:
Same here, will an alloc always have a DeploymentStatus?

Reply:
Also this logic is pretty confusing- this will match no matter what, correct? Should you assert that the status is actually set as opposed to the status of the health?

Reply:
Essentially it is checking the two possible set states, since the value is a pointer, it is using the two helpers

Reply:
Yeah, a `HasHealthSet()` helper method that encapsulated this logic would make it much more clear what was going on and clean up places where we do an equivalent check with slightly different code: `alloc.DeploymentStatus != nil && alloc.DeploymentStatus.Healthy != nil`

Might sneak that into a followup PR.